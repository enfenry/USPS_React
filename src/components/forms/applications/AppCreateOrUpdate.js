import React from 'react';
import { Button, Form, FormGroup } from 'reactstrap';
import { reduxForm, Field, formValueSelector } from 'redux-form';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ADDRESS_CHANGE, MAIL_FORWARDING, PACKAGE_SUBMISSION } from '../../../constants/applicationTypes';
import validate from './AppValidate';
import { renderField } from '../formUtils';

 function getApplicationTypeName(applicationType) {
  switch (applicationType) {
    case ADDRESS_CHANGE:
      return "Address Change";
    case MAIL_FORWARDING:
      return "Mail Forwarding";
    case PACKAGE_SUBMISSION:
      return "Package Submission";
    default:
      return applicationType;
  }
}

let AppCreateOrUpdate = props => {
  const { handleSubmit, handleHide, appTypeValue, customers, addresses, products } = props;

  const isPackageSubmission = parseInt(appTypeValue) === PACKAGE_SUBMISSION;
  const shippingSpeeds = products.filter(product => product.hierarchypath === "USPS\\Shipping Speed");

  const renderOptions = (array, value, display) => {
    return array.map(el => {
      return (
        <option key={el[value]} value={el[value]}>{el[display]}</option>
      )
    })
  }

  const filterProducts = () => {
    return products.filter(product => product.hierarchypath === `USPS\\${(getApplicationTypeName(parseInt(appTypeValue)))}`);
  }

  return (
    <Form onSubmit={handleSubmit} className="form">
      <FormGroup className="field">
        <div className="control">
          <Field name="ss_name" component={renderField} type="text" label="Name" />
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="ss_applicationtype" component={renderField} type="select"
            label="Application Type">
            <option value={null}></option>
            <option value={ADDRESS_CHANGE}>Address Change</option>
            <option value={MAIL_FORWARDING}>Mail Forwarding</option>
            <option value={PACKAGE_SUBMISSION}>Package Submission</option>
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_product_value" component={renderField} type="select"
            label="Product">
            <option value={null}></option>
            {renderOptions(filterProducts(), "productid", "name")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field show={isPackageSubmission} name="_ss_shippingspeed_value" component={renderField} type="select"
            label="Shipping Speed">
            <option value={null}></option>
            {renderOptions(shippingSpeeds, "productid", "name")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_customer_value" component={renderField} type="select"
            label="Customer">
            <option value={null}></option>
            {renderOptions(customers, "contactid", "fullname")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_destinationaddress_value" component={renderField} type="select"
            label="Destination Address" >
            <option value={null}></option>
            {renderOptions(addresses, "ss_customaddressid", "ss_name")}
          </Field>
        </div>
      </FormGroup>

      <div className="field">
        <div className="control">
          <Button className="button is-link" color="success" onClick={handleSubmit}>Submit</Button>{' '}
          <Button color="secondary" onClick={handleHide}>Cancel</Button>
        </div>
      </div>
    </Form>
  );
};

AppCreateOrUpdate.propTypes = {
  handleSubmit: PropTypes.func,
  handleHide: PropTypes.func,
  initialValues: PropTypes.object,
  customers: PropTypes.arrayOf(PropTypes.object),
  addresses: PropTypes.arrayOf(PropTypes.object),
  products: PropTypes.arrayOf(PropTypes.object),
  appTypeValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

AppCreateOrUpdate = reduxForm({
  form: 'AppCreateOrUpdate',
  validate,
  onSubmitSuccess: (result, dispatch, props) => {
    props.handleHide();
  }
})(AppCreateOrUpdate);

function mapStateToProps(state) {
  const selector = formValueSelector('AppCreateOrUpdate');
  const appTypeValue = selector(state, 'ss_applicationtype')

  return {
    appTypeValue,
    customers: state.customersReducer.customers,
    addresses: state.addressesReducer.addresses,
    products: state.productsReducer.products,
  }
}

export default connect(
  mapStateToProps
)(AppCreateOrUpdate);