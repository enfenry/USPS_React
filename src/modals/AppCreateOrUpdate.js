import React from 'react';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import { reduxForm, Field, formValueSelector } from 'redux-form';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getApplicationTypeName } from '../components/ApplicationsRender';
import { ADDRESS_CHANGE, MAIL_FORWARDING, PACKAGE_SUBMISSION } from '../constants/applicationTypes';

let AppCreateOrUpdate = props => {
  const { handleSubmit, handleHide, appTypeValue, customers, addresses, products } = props;

  const isPackageSubmission = parseInt(appTypeValue) === PACKAGE_SUBMISSION;
  const shippingSpeeds = products.filter(product => product.hierarchypath === "USPS\\Shipping Speed");

  const renderOptions = (array, value, display) => {
    return array.map(el => {
      return (
        <option key={el[value]} value={el[value]}>{el[display]}</option>
      )
    })
  }

  const filterProducts = () => {
    return products.filter(product => product.hierarchypath === `USPS\\${(getApplicationTypeName(parseInt(appTypeValue)))}`);
  }

  return (
    <Form onSubmit={handleSubmit} className="form">
      <FormGroup className="field">
        <div className="control">
          <Field name="ss_name" component={renderField} type="text" label="Name" />
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="ss_applicationtype" component={renderField} type="select"
            label="Application Type">
            <option value={null}></option>
            <option value={ADDRESS_CHANGE}>Address Change</option>
            <option value={MAIL_FORWARDING}>Mail Forwarding</option>
            <option value={PACKAGE_SUBMISSION}>Package Submission</option>
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_product_value" component={renderField} type="select"
            label="Product">
            <option value={null}></option>
            {renderOptions(filterProducts(), "productid", "name")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field show={isPackageSubmission} name="_ss_shippingspeed_value" component={renderField} type="select"
            label="Shipping Speed">
            <option value={null}></option>
            {renderOptions(shippingSpeeds, "productid", "name")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_customer_value" component={renderField} type="select"
            label="Customer">
            <option value={null}></option>
            {renderOptions(customers, "contactid", "fullname")}
          </Field>
        </div>
      </FormGroup>

      <FormGroup className="field">
        <div className="control">
          <Field name="_ss_destinationaddress_value" component={renderField} type="select"
            label="Destination Address" >
            <option value={null}></option>
            {renderOptions(addresses, "ss_customaddressid", "ss_name")}
          </Field>
        </div>
      </FormGroup>


      <div className="field">
        <div className="control">
          <Button className="button is-link" color="success" onClick={handleSubmit}>Submit</Button>{' '}
          <Button color="secondary" onClick={handleHide}>Cancel</Button>
        </div>
      </div>
    </Form>
  );
};


const validate = (values) => {
  const errors = {};

  const checkRequiredField = (fieldName) => {
    const currentVal = values[fieldName];

    if (!(currentVal) || currentVal === null) {
      errors[fieldName] = 'Required';
    }
    else if (typeof currentVal === 'string') {

      if (currentVal.trim() === '') {
        errors[fieldName] = 'Cannot be empty';
      }
    }
  }

  const isPackageSubmission = (parseInt(values.ss_applicationtype)) === PACKAGE_SUBMISSION;

  checkRequiredField('ss_name');
  checkRequiredField('ss_applicationtype');
  checkRequiredField('_ss_product_value');
  if (isPackageSubmission) {
    checkRequiredField('_ss_shippingspeed_value');
  }
  checkRequiredField('_ss_customer_value');
  checkRequiredField('_ss_destinationaddress_value');

  return errors;
};

const renderField = (props) => {

  const { input, label, type, show, meta: { touched, error, warning } } = props;

  return (
    <div style={{ display: show ? 'block' : 'none' }} >
      <div className="control">
        <Label className="field">{label}</Label>
        <Input className="input" {...input} type={type}>
          {props.children}
        </Input>
        {touched && ((error && <span style={{ color: 'red' }}>{error}</span>) || (warning && <span style={{ color: 'orange' }}>{warning}</span>))}
      </div>
    </div>
  )
}

renderField.defaultProps = {
  show: true
}

renderField.propTypes = {
  input: PropTypes.object,
  label: PropTypes.string,
  type: PropTypes.string,
  show: PropTypes.bool,
  meta: PropTypes.object,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

AppCreateOrUpdate.propTypes = {
  handleSubmit: PropTypes.func,
  handleHide: PropTypes.func,
  initialValues: PropTypes.object,
  customers: PropTypes.arrayOf(PropTypes.object),
  addresses: PropTypes.arrayOf(PropTypes.object),
  products: PropTypes.arrayOf(PropTypes.object),
  appTypeValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

AppCreateOrUpdate = reduxForm({
  form: 'AppCreateOrUpdate',
  validate,
  onSubmitSuccess: (result, dispatch, props) => {
    props.handleHide();
  }
})(AppCreateOrUpdate);

function mapStateToProps(state) {
  const selector = formValueSelector('AppCreateOrUpdate');
  const appTypeValue = selector(state, 'ss_applicationtype')

  return {
    appTypeValue,
    customers: state.customersReducer.customers,
    addresses: state.addressesReducer.addresses,
    products: state.productsReducer.products,
  }
}

export default connect(
  mapStateToProps
)(AppCreateOrUpdate);