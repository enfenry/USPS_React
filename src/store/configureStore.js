import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';

import { createLogger } from 'redux-logger'
import DevTools from '../containers/DevTools';

function configureStore(initialState) {

  const logger = createLogger()

  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    logger
  ];

  const store = createStore(
    rootReducer,
    initialState,
    compose(
      applyMiddleware(...middlewares),
      DevTools.instrument()
    )
  );

  return store;
}

export default configureStore;
