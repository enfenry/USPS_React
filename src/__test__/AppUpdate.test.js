import React from 'react';
import AppCreateOrUpdate from '../modals/AppCreateOrUpdate';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

test('AppCreateOrUpdate renders correctly', () => {

    const store = mockStore({
        productsReducer: { products: [] },
        ordersReducer: { orders: [] },
        customersReducer: {
            customers: [
                {
                    contactid: "A",
                    fullname: "Alice Astronaut"
                },
                {
                    contactid: "B",
                    fullname: "Bobby Butcher"
                },
                {
                    contactid: "C",
                    fullname: "Charlie Cobbler"
                }
            ],
        },
        addressesReducer: { addresses: [] },
    });

    const component = renderer.create(
        <Provider store={store}>
            <AppCreateOrUpdate
                data={{
                    _ss_customer_value: "B"
                }}
            />
        </Provider>
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});