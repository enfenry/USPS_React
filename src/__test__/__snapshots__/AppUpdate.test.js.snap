// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppUpdate renders correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Name
          </label>
          <input
            className="input form-control"
            name="ss_name"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Application Type
          </label>
          <select
            className="input form-control"
            name="ss_applicationtype"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
          >
            <option
              value="Address Change"
            >
              Address Change
            </option>
            <option
              value="Package Submission"
            >
              Package Submission
            </option>
            <option
              value="Mail Forwarding"
            >
              Mail Forwarding
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Product
          </label>
          <select
            className="input form-control"
            name="_ss_product_value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
          >
            <option
              value={null}
            />
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Shipping Speed
          </label>
          <select
            className="input form-control"
            name="_ss_shippingspeed_value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
          >
            <option
              value={null}
            />
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Customer
          </label>
          <select
            className="input form-control"
            defaultValue="B"
            name="_ss_customer_value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
          >
            <option
              value={null}
            />
            <option
              value="A"
            >
              Alice Astronaut
            </option>
            <option
              value="B"
            >
              Bobby Butcher
            </option>
            <option
              value="C"
            >
              Charlie Cobbler
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field form-group"
  >
    <div
      className="control"
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <div
          className="control"
        >
          <label
            className="field"
          >
            Destination Address
          </label>
          <select
            className="input form-control"
            name="_ss_destinationaddress_value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
          >
            <option
              value={null}
            />
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field"
  >
    <div
      className="control"
    >
      <button
        aria-label={null}
        className="button is-link btn btn-success"
        onClick={[Function]}
      >
        Submit
      </button>
       
      <button
        aria-label={null}
        className="btn btn-secondary"
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </div>
</form>
`;
